File: host-libs/android/src/main/kotlin/io/parity/truapi/TrUAPIHost.kt
// TrUAPIHost - Android host adapter.
//
// The Rust core (compiled to `libtruapi_server.so` and surfaced via UniFFI in
// `src/main/kotlin/generated/uniffi/truapi_server/truapi_server.kt`) owns the
// wire protocol, request routing, subscription lifecycle, and platform trait
// dispatch.
//
// This file exposes two things on top of the generated bindings:
//
//   * `HostBridge` - a Kotlin-friendly callback interface the embedding app
//     implements. It splits device and remote permissions, mirroring the
//     `Permissions` platform trait in the Rust core.
//   * `WebViewTransport` - a thin byte transport that forwards opaque wire
//     bytes between a `WebView` and the Rust core. Bytes traverse the
//     `JavascriptInterface` boundary as base64 because the bridge cannot
//     carry binary types directly.
//
// The transport is independent of the core: tests can stand up a
// `WebViewTransport` against a non-UniFFI stub by implementing `CoreInbound`.

package io.parity.truapi

import android.util.Base64
import android.webkit.JavascriptInterface
import android.webkit.WebView
import uniffi.truapi_server.HostCallbacks
import uniffi.truapi_server.HostNavigateRejection
import uniffi.truapi_server.HostRejection
import uniffi.truapi_server.HostStorageException
import uniffi.truapi_server.NativeTrUApiCore
import uniffi.truapi_server.WsBridgeEndpoint
import uniffi.truapi_server.WsBridgeStartException

/** Package metadata. */
object TrUAPIHost {
    const val VERSION = "0.1.0"
}

/**
 * Storage backend the host provides to the Rust core. Throws
 * [HostStorageException] to signal quota exhaustion or unknown failure; the
 * core maps both onto the v0.1 `HostLocalStorageReadError` wire shape.
 */
interface HostStorage {
    @Throws(HostStorageException::class)
    fun read(key: String): ByteArray?

    @Throws(HostStorageException::class)
    fun write(key: String, value: ByteArray)

    @Throws(HostStorageException::class)
    fun clear(key: String)
}

/**
 * Host-side callback bundle that the Rust core invokes for capabilities the
 * native shell owns. The interface mirrors the underlying UniFFI surface
 * but keeps the permission split explicit:
 *
 *   * [devicePermission] handles camera / mic / push prompts and similar
 *     OS-scoped grants. `request` is a SCALE-encoded
 *     `v01::HostDevicePermissionRequest`.
 *   * [remotePermission] handles per-product capability bundles requested
 *     by the application running inside the WebView. `request` is a
 *     SCALE-encoded `v01::RemotePermissionRequest`.
 *
 * Embedders typically wire the SCALE payloads through the generated
 * `@parity/truapi` client running on the JS side for UI rendering, then
 * report the user's decision as a `Boolean`.
 */
interface HostBridge {
    /** Lifecycle logger. Marker is a stable slug, detail is free-form. */
    fun onCoreLog(marker: String, detail: String) {}

    /** Forward an outbound SCALE-encoded protocol frame to the product. */
    fun onCoreResponse(frame: ByteArray)

    /** Open a URL in the system browser. */
    @Throws(HostNavigateRejection::class)
    fun navigateTo(url: String)

    /** Deliver a push notification (SCALE-encoded `HostPushNotificationRequest`). */
    @Throws(HostRejection::class)
    fun pushNotification(payload: ByteArray)

    /** Prompt for a device-level permission. Returns whether it was granted. */
    @Throws(HostRejection::class)
    fun devicePermission(request: ByteArray): Boolean

    /** Prompt for a remote (product-scoped) permission bundle. */
    @Throws(HostRejection::class)
    fun remotePermission(request: ByteArray): Boolean

    /** Answer a feature-support query. */
    @Throws(HostRejection::class)
    fun featureSupported(request: ByteArray): Boolean

    /** Scoped key-value storage for the Rust core. */
    val storage: HostStorage
}

/**
 * Adapter from the public [HostBridge] surface to the generated UniFFI
 * [HostCallbacks] interface. Keeps the public API stable even if uniffi-bindgen
 * renames generated symbols.
 */
private class HostCallbackAdapter(private val bridge: HostBridge) : HostCallbacks {
    override fun onCoreLog(marker: String, detail: String) =
        bridge.onCoreLog(marker, detail)

    override fun onCoreResponse(frame: ByteArray) =
        bridge.onCoreResponse(frame)

    override fun navigateTo(url: String) =
        bridge.navigateTo(url)

    override fun pushNotification(payload: ByteArray) =
        bridge.pushNotification(payload)

    override fun devicePermission(request: ByteArray): Boolean =
        bridge.devicePermission(request)

    override fun remotePermission(request: ByteArray): Boolean =
        bridge.remotePermission(request)

    override fun featureSupported(request: ByteArray): Boolean =
        bridge.featureSupported(request)

    override fun localStorageRead(key: String): ByteArray? =
        bridge.storage.read(key)

    override fun localStorageWrite(key: String, value: ByteArray) =
        bridge.storage.write(key, value)

    override fun localStorageClear(key: String) =
        bridge.storage.clear(key)
}

/**
 * Sink for opaque wire frames coming from the WebView. The Rust core is the
 * typical implementor (via [TrUAPIHostCore]); tests may use a stub.
 */
fun interface CoreInbound {
    fun receiveFromProduct(frame: ByteArray)
}

/**
 * Owning wrapper around the Rust-backed [NativeTrUApiCore]. Implements
 * [CoreInbound] so a [WebViewTransport] can deliver inbound frames directly,
 * and exposes session and WS bridge controls.
 *
 * The wrapper holds a strong reference to the bridge so the JNA callback
 * registration stays alive for the lifetime of the core.
 */
class TrUAPIHostCore(bridge: HostBridge) : CoreInbound, AutoCloseable {
    @Suppress("unused") // retained to keep JNA callbacks alive
    private val callbackRetainer: HostCallbacks = HostCallbackAdapter(bridge)
    private val inner: NativeTrUApiCore = NativeTrUApiCore(callbackRetainer)

    override fun receiveFromProduct(frame: ByteArray): Unit {
        inner.receiveFromProduct(frame)
    }

    /** Set the currently-paired session. `pubkey` must be exactly 32 bytes. */
    fun setActiveSession(
        pubkey: ByteArray,
        liteUsername: String? = null,
        fullUsername: String? = null,
    ): Boolean = inner.setActiveSession(pubkey, liteUsername, fullUsername)

    /** Drop the currently-paired session. */
    fun clearActiveSession() {
        inner.clearActiveSession()
    }

    /** Start the localhost WebSocket bridge (requires the `ws-bridge` feature in the cdylib). */
    @Throws(WsBridgeStartException::class)
    fun startWsBridge(bindPort: UShort = 0u): WsBridgeEndpoint =
        inner.startWsBridge(bindPort)

    /** Stop the localhost WebSocket bridge (if running). */
    fun stopWsBridge() {
        inner.stopWsBridge()
    }

    /** Smoke-test helper: returns a SCALE-encoded `feature_supported` request frame. */
    fun debugSmokeFeatureRequestFrame(): ByteArray =
        inner.debugSmokeFeatureRequestFrame()

    override fun close() {
        inner.close()
    }
}

/**
 * Wraps a [WebView] and forwards opaque wire bytes between JS and [core].
 * Attach with [attach] before loading the page so the JS shim is installed.
 */
class WebViewTransport(
    private val webView: WebView,
    private val core: CoreInbound,
    private val callbackName: String = "__trUApiReceive",
    private val interfaceName: String = "TrUApi",
) {
    fun attach() {
        webView.addJavascriptInterface(JsInterface(), interfaceName)
    }

    fun detach() {
        webView.removeJavascriptInterface(interfaceName)
    }

    /**
     * JS bootstrap to inject at document start so the page exposes a
     * `window.trUApi` byte-pipe matching the JS host adapter shape.
     */
    val bootstrapScript: String = """
        (function() {
          var listeners = [];
          window.$callbackName = function(b64) {
            try {
              var bin = atob(b64);
              var bytes = new Uint8Array(bin.length);
              for (var i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
              listeners.forEach(function(l) { l(bytes); });
            } catch (e) { console.error('trUApi recv error', e); }
          };
          function toB64(u8) {
            var s = '';
            for (var i = 0; i < u8.length; i++) s += String.fromCharCode(u8[i]);
            return btoa(s);
          }
          window.trUApi = {
            postMessage: function(bytes) {
              window.$interfaceName.postMessage(toB64(bytes));
            },
            subscribe: function(cb) {
              listeners.push(cb);
              return function() {
                var i = listeners.indexOf(cb);
                if (i >= 0) listeners.splice(i, 1);
              };
            }
          };
          window.dispatchEvent(new Event('truapi-native-ready'));
        })();
    """.trimIndent()

    /** Called when the core has bytes to push to the product app. */
    fun sendToProduct(frame: ByteArray) {
        val b64 = Base64.encodeToString(frame, Base64.NO_WRAP)
        val js = "window.$callbackName && window.$callbackName('$b64')"
        webView.post { webView.evaluateJavascript(js, null) }
    }

    private inner class JsInterface {
        @JavascriptInterface
        fun postMessage(b64: String) {
            val frame = try {
                Base64.decode(b64, Base64.NO_WRAP)
            } catch (_: IllegalArgumentException) {
                return
            }
            core.receiveFromProduct(frame)
        }
    }
}

File: ios/TrUAPIHost/Sources/TrUAPIHost/TrUAPIHost.swift
// TrUAPIHost - iOS host adapter.
//
// The Rust core (compiled to `libtruapi_server`, surfaced through UniFFI in
// the sibling `truapi_server.swift` file) owns wire decoding, request
// routing, subscription lifecycle, and platform trait dispatch.
//
// This file layers two things on top of the generated bindings:
//
//   * `HostBridge` - a Swift-friendly callback bundle the embedding app
//     implements. It splits device and remote permissions, mirroring the
//     `Permissions` platform trait in the Rust core.
//   * `WebViewTransport` - a base64-over-`WKScriptMessageHandler` byte pipe
//     between a `WKWebView` and any `CoreInbound`.
//
// `LocalhostBridgeBootstrap` is retained from earlier iOS shells for hosts
// that prefer the localhost WebSocket bridge over the direct WK script
// bridge.

import Foundation
import WebKit

/// Package metadata.
public enum TrUAPIHost {
    public static let version = "0.1.0"
}

/// Bootstrap helper for the native localhost WebSocket bridge that the Rust
/// core can stand up via `NativeTrUApiCore.startWsBridge(bindPort:)` when
/// the cdylib is built with the `ws-bridge` feature.
public enum LocalhostBridgeBootstrap {
    /// Returns a `<script>`-injectable snippet that publishes the endpoint
    /// metadata on `window.__truapi_localhost` and fires a `truapi-native-ready`
    /// event. The product client reads this and dials back in.
    public static func script(port: UInt16, token: String) -> String {
        let url = "ws://127.0.0.1:\(port)/?t=\(token)"
        let safeUrl = escapeJavaScriptString(url)
        let safeToken = escapeJavaScriptString(token)
        return """
        (function() {
          window.__truapi_localhost = { url: '\(safeUrl)', token: '\(safeToken)' };
          window.dispatchEvent(new Event('truapi-native-ready'));
        })();
        """
    }

    private static func escapeJavaScriptString(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
    }
}

/// Storage backend the host provides to the Rust core. Throwing closures
/// can surface quota or unknown failures by raising `HostStorageError`
/// (defined in the generated bindings).
public protocol HostStorageBackend: AnyObject, Sendable {
    func read(key: String) throws -> Data?
    func write(key: String, value: Data) throws
    func clear(key: String) throws
}

/// Host-side callback bundle that the Rust core invokes for capabilities the
/// native shell owns. The permission split mirrors the Rust `Permissions`
/// trait:
///
///   * ``devicePermission(request:)`` handles OS-scoped grants (camera,
///     mic, location). `request` is a SCALE-encoded
///     `v01::HostDevicePermissionRequest`.
///   * ``remotePermission(request:)`` handles per-product capability
///     bundles. `request` is a SCALE-encoded `v01::RemotePermissionRequest`.
///
/// Embedders typically forward the SCALE payloads through the
/// `@parity/truapi` JS client for UI prompts, then return the boolean
/// granted flag.
public protocol HostBridge: AnyObject, Sendable {
    /// Lifecycle logger. Marker is a stable slug, detail is free-form.
    func onCoreLog(marker: String, detail: String)

    /// Forward an outbound SCALE-encoded protocol frame to the product.
    func onCoreResponse(frame: Data)

    /// Open a URL in the system browser.
    func navigateTo(url: String) throws

    /// Deliver a push notification (SCALE-encoded `HostPushNotificationRequest`).
    func pushNotification(payload: Data) throws

    /// Prompt for a device-level permission. Returns the granted flag.
    func devicePermission(request: Data) throws -> Bool

    /// Prompt for a remote (product-scoped) permission bundle.
    func remotePermission(request: Data) throws -> Bool

    /// Answer a feature-support query.
    func featureSupported(request: Data) throws -> Bool

    /// Scoped key-value storage for the Rust core.
    var storage: HostStorageBackend { get }
}

public extension HostBridge {
    /// Default no-op logger. Override to plumb into your logging framework.
    func onCoreLog(marker: String, detail: String) {}
}

/// Adapter that bridges the public `HostBridge` to the generated UniFFI
/// `HostCallbacks` protocol. Kept private so the generated names never
/// leak into consumers.
private final class HostCallbackAdapter: HostCallbacks, @unchecked Sendable {
    private let bridge: HostBridge

    init(bridge: HostBridge) {
        self.bridge = bridge
    }

    func onCoreLog(marker: String, detail: String) {
        bridge.onCoreLog(marker: marker, detail: detail)
    }

    func onCoreResponse(frame: Data) {
        bridge.onCoreResponse(frame: frame)
    }

    func navigateTo(url: String) throws {
        try bridge.navigateTo(url: url)
    }

    func pushNotification(payload: Data) throws {
        try bridge.pushNotification(payload: payload)
    }

    func devicePermission(request: Data) throws -> Bool {
        try bridge.devicePermission(request: request)
    }

    func remotePermission(request: Data) throws -> Bool {
        try bridge.remotePermission(request: request)
    }

    func featureSupported(request: Data) throws -> Bool {
        try bridge.featureSupported(request: request)
    }

    func localStorageRead(key: String) throws -> Data? {
        try bridge.storage.read(key: key)
    }

    func localStorageWrite(key: String, value: Data) throws {
        try bridge.storage.write(key: key, value: value)
    }

    func localStorageClear(key: String) throws {
        try bridge.storage.clear(key: key)
    }
}

/// Sink for opaque wire frames coming from the WebView. The Rust core
/// (via `TrUAPIHostCore`) is the typical implementor; tests may use a stub.
public protocol CoreInbound: AnyObject {
    func receiveFromProduct(_ frame: Data)
}

/// Owning wrapper around the Rust-backed `NativeTrUApiCore`. Implements
/// `CoreInbound` so a `WebViewTransport` can hand inbound frames over
/// directly, and exposes session and WS bridge controls.
///
/// Holds a strong reference to the bridge adapter so the UniFFI callback
/// vtable stays valid for the lifetime of the core.
public final class TrUAPIHostCore: CoreInbound {
    private let inner: NativeTrUApiCore
    // Retained so the UniFFI callback vtable stays valid for the lifetime
    // of `inner`. Not read directly; the suppression keeps -Wunused happy.
    private let callbackRetainer: HostCallbacks

    public init(bridge: HostBridge) {
        let adapter = HostCallbackAdapter(bridge: bridge)
        self.callbackRetainer = adapter
        self.inner = NativeTrUApiCore(callbacks: adapter)
        _ = self.callbackRetainer
    }

    public func receiveFromProduct(_ frame: Data) {
        _ = inner.receiveFromProduct(frame: frame)
    }

    /// Set the currently-paired session. `pubkey` must be exactly 32 bytes.
    @discardableResult
    public func setActiveSession(
        pubkey: Data,
        liteUsername: String? = nil,
        fullUsername: String? = nil
    ) -> Bool {
        inner.setActiveSession(pubkey: pubkey, liteUsername: liteUsername, fullUsername: fullUsername)
    }

    /// Drop the currently-paired session.
    public func clearActiveSession() {
        inner.clearActiveSession()
    }

    /// Start the localhost WebSocket bridge.
    /// Requires the `ws-bridge` feature in the cdylib.
    public func startWsBridge(bindPort: UInt16 = 0) throws -> WsBridgeEndpoint {
        try inner.startWsBridge(bindPort: bindPort)
    }

    /// Stop the localhost WebSocket bridge (if running).
    public func stopWsBridge() {
        inner.stopWsBridge()
    }

    /// Smoke-test helper: returns a SCALE-encoded `feature_supported`
    /// request frame so the iOS shell can verify the wire path.
    public func debugSmokeFeatureRequestFrame() -> Data {
        inner.debugSmokeFeatureRequestFrame()
    }
}

/// The iOS-side byte transport. Wraps a `WKWebView`; forwards bytes between
/// the JS bridge and a `CoreInbound` (typically a `TrUAPIHostCore`).
@MainActor
public final class WebViewTransport: NSObject, WKScriptMessageHandler {
    private weak var webView: WKWebView?
    private weak var core: AnyObject?
    private let coreSend: (Data) -> Void
    private let callbackName: String
    private let messageName: String

    public init(
        webView: WKWebView,
        core: CoreInbound,
        callbackName: String = "__trUApiReceive",
        messageName: String = "trUApi"
    ) {
        self.webView = webView
        self.core = core
        self.coreSend = { [weak core] frame in core?.receiveFromProduct(frame) }
        self.callbackName = callbackName
        self.messageName = messageName
        super.init()
    }

    /// JS bootstrap to inject as a `WKUserScript` so the page exposes a
    /// `window.trUApi` byte-pipe matching the JS host adapter shape.
    public var bootstrapScript: String {
        return """
        (function() {
          var listeners = [];
          window.\(callbackName) = function(b64) {
            try {
              var bin = atob(b64);
              var bytes = new Uint8Array(bin.length);
              for (var i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
              listeners.forEach(function(l) { l(bytes); });
            } catch (e) { console.error('trUApi recv error', e); }
          };
          function toB64(u8) {
            var s = '';
            for (var i = 0; i < u8.length; i++) s += String.fromCharCode(u8[i]);
            return btoa(s);
          }
          window.trUApi = {
            postMessage: function(bytes) {
              window.webkit.messageHandlers.\(messageName).postMessage(toB64(bytes));
            },
            subscribe: function(cb) {
              listeners.push(cb);
              return function() {
                var i = listeners.indexOf(cb);
                if (i >= 0) listeners.splice(i, 1);
              };
            }
          };
          window.dispatchEvent(new Event('truapi-native-ready'));
        })();
        """
    }

    public func attach(to controller: WKUserContentController) {
        controller.add(self, name: messageName)
        let script = WKUserScript(source: bootstrapScript, injectionTime: .atDocumentStart, forMainFrameOnly: true)
        controller.addUserScript(script)
    }

    public func detach(from controller: WKUserContentController) {
        controller.removeScriptMessageHandler(forName: messageName)
    }

    /// Called by the host (typically from `HostBridge.onCoreResponse`) when
    /// the core has bytes to push back into the product app.
    public func sendToProduct(_ frame: Data) {
        guard let webView else { return }
        let b64 = frame.base64EncodedString()
        let js = "window.\(callbackName) && window.\(callbackName)('\(b64)')"
        webView.evaluateJavaScript(js, completionHandler: nil)
    }

    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let b64 = message.body as? String,
              let data = Data(base64Encoded: b64) else { return }
        coreSend(data)
    }
}
